File: middleLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        int middle = 1;
        ListNode *temp = head;
        
        
        //calculate the total nodes
        while(temp -> next != NULL){
            middle ++;
            temp = temp -> next;
        }
        
        middle = (middle / 2) + 1;
        temp = head;
        
        //search for the middle node
        for(int i = 1; i < middle; i++){
            temp = temp -> next;
        }
        
        return temp;
    }
};

File: progress_sheet/middleLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:

        # start from head
        slow = head
        fast = head

        # move second pointer twice as fast as first pointer
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # fast.next => to handle odd elements
        # fast => to handle even elements
        # interchangine fast and fast.next doesn't work if it's even
         
        return slow
